Node edge type storage format

e-mcginnis opened the pull request:
## Creates the following mappings in libtsuba
```
node_entity_type_id_array
edge_entity_type_id_array
```
which map from node/edge id to EntityTypeID


```
node_entity_type_id_dictionary
edge_entity_type_id_dictionary
```
which map from EntityTypeID to set of EntityTypeIDs that define it
Atmoic EntityTypeIDs map to a set containing themselves
Intersection EntityTypeIDs map to a set of EntityTypeIDs the define it


```
node_entity_type_id_name
edge_entity_type_id_name
```
which maps from Atomic EntityTypeID to string name for the EntityTypeID


### deprecates node/edge type ids and names stored in the properties 


### Adds storage_format_version to the RDG part header to track changes to its structure.
We then use this to determine if we should load EntityTypeIDs from the 6 structures described above. If it is not present, then we have an old graph and should instead load node types from properties. 


# Additional changes
Fix python wrapper for PropertyGraph

Add Equal and Diff functions for EntityTypeManager

update PG:Equals and PG:Diff functions to include new type storage
structures

add TESTING ONLY header to Equals/Diff functions

Fix EntityTypeIDManager Init off by one
When initializing the EntityTypeIDManager we added sets for id = 0
aka kUnknownType, but we didn't properly initialize the set
the set for EntityTypeID 0 should contain itself
Added a name for kUnknownType so we can just use the proper
AddAtomicEntityType

enforce topo_size = entity_type_ids_size when converting properties to
types

Added test for storing and loading types (Scenario 1)
Added test for storing and loading composite types (Scenario 2)
Added test for loading existing rdg, converting it, storing it back
(Scenario 3)

Co-authored-by: Evan McGinnis <evan@katanagraph.com>
Co-authored-by: Giorgi Lomia <lomiag@berea.edu>
Co-authored-by: M. Amber Hassaan <amber_474@yahoo.com>

roshandathathri commented:
Note that this PR is ready to be reviewed but ctest will fail for personcityuniversity.q. I'll be updating "inputs" separately to remove that file from ctest.

l-hoang commented on .github/workflows/cpp.yaml:
```diff
@@ -166,7 +166,7 @@ jobs:
         GRAPH_FILE="$HOME/.cache/graph/inputs.tar.gz"
         GRAPH_LOCATION="$HOME/.cache/graph/inputs"
-        INPUT_URL="https://katana-ci-public.s3.us-east-1.amazonaws.com/inputs/katana-inputs-v2.1.tar.gz"
+        INPUT_URL="https://katana-ci-public.s3.us-east-1.amazonaws.com/inputs/katana-inputs-v3.tar.gz"
```
I would think removing one line would really only push it to 2.2 and not v3. :P

roshandathathri replied:
No minor versions for inputs. That's what @ddn0 suggested and I agree.

roshandathathri commented:
Inputs have been updated, so everything should pass now (hopefully).

l-hoang commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -408,40 +398,22 @@ class SubgraphQueryMiner {
-        exploreCandidates<true, DFS, printEmbeddings>(
-            d_vertex, qeData.label, index, emb, neighbors, numInNeighbors,
-            out_queue);
+    // d_vertex is incoming neighbor
```
This comment is misleading, no? You can only check  outgoing neighbors.

roshandathathri replied:
I understand that it can be confusing but the comment is precise. "neighbor" here refers to the neighbor of the query vertex being considered. d_vertex is an incoming neighbor of that query vertex. The outgoing neighbors of d_vertex will be candidates (to match) for that query vertex.

l-hoang commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -346,26 +343,19 @@ class SubgraphQueryMiner {
                          unsigned neighborIndex, const QueryEmbeddingType& emb,
                          NeighborsTy& neighbors, unsigned numInNeighbors,
                          QueryEmbeddingQueueType& out_queue) {
     unsigned n = emb.size();
```
For the time being it may be worth it to assert that numInNeighbors is 0.

roshandathathri replied:
You mean assert that it is NOT 0? Line 297 is doing that - it is appropriate there because the code following the assertion assumes that.

To be clear, there can be both incoming and outgoing neighbors of the query vertex. This is handled in toAdd() - check line 157. There should be at least one incoming neighbor of the query vertex, so that we can iterate over the outgoing edges of that neighbor to find candidates (otherwise, we would have to consider every vertex as a candidate).

l-hoang commented on libquery/src/SubgraphQuery.cpp:
```diff
@@ -304,7 +305,7 @@ class SubgraphQueryMiner {
     unsigned index       = neighbors.size();
     size_t numCandidates = dataGraph.size(); // conservative upper limit
-    for (unsigned i = 0; i < neighbors.size(); ++i) {
+    for (unsigned i = 0; i < numInNeighbors; ++i) {
```
Shouldn't numInNeighbors always be 0? IIRC it was just an index into which neighbors were in-neighbors, and if you don't have the incoming edges you would always have 0.

roshandathathri replied:
No. It should NOT be 0. Hopefully, this should be clarified by the comments above.

roshandathathri commented:
@ddn0 any idea why it's taking around an hour to build "Shared" on macOS?

e-mcginnis committed:
EntityTypeID management in PG and libtsuba

Fix python wrapper for PropertyGraph

Add Equal and Diff functions for EntityTypeManager

update PG:Equals and PG:Diff functions to include new type storeage
structures

add TESTING ONLY header to Equals/Diff functions

Fix EntityTypeIDManager Init off by one
When initializing the EntityTypeIDManager we added sets for id = 0
aka kUnknownType, but we didn't properly initialize the set
the set for EntityTypeID 0 should contain itself
Added a name for kUnknownType so we can just use the proper
AddAtomicEntityType

enforce topo_size = entity_type_ids_size when converting properties to
types

Added test for storing and loading types (Scenario 1)
Added test for storing and loading composite types (Scenario 2)
Added test for loading existing rdg, converting it, storing it back
(Scenario 3)

Co-authored-by: Evan McGinnis <evan@katanagraph.com>
Co-authored-by: Giorgi Lomia <lomiag@berea.edu>"
Co-authored-by: M. Amber Hassaan <amber_474@yahoo.com>

tylershunt commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -539,6 +539,16 @@ class KATANA_EXPORT PropertyGraph {
+  /// \returns a pointer to the underlying memory of the Edge type ID array
+  TypeSetID* GetEdgeTypeSetIDArrayData() const {
+    return edge_type_set_id_.data();
+  }
```
Would normally prefer just a const reference to `edge_type_set_id`.

e-mcginnis replied:
@GioLomia

amberhassaan replied:
I actually prefer not returning references to internal data structures, such as `NUMAArray` to prevent accidental misuse.

roshandathathri replied:
Agree with Tyler

e-mcginnis commented on libtsuba/src/RDGPartHeader.h:
```diff
@@ -271,7 +338,38 @@ class KATANA_EXPORT RDGPartHeader {
 
+  // tracks changes to json on disk structure of the PartitionHeader
+  // current one is defined by latest_storage_format_version_
+  uint32_t storage_format_version_;
```
@tylershunt just an FYI for eventual review. 

I added `storage_format_version` so that we can track json layout changes to the RDGPartHeader. 
This made it easy to differentiate between part headers that should have the new EntityTypeID structures vs ones that don't in the RDGPartHeader `from_json`, and the RDG `Make` and `Validate` funcitons
defaulted to using `1` for old version, `2` for new.
on store, we always write the latest

tylershunt replied:
Sounds good.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -237,6 +237,43 @@ class KATANA_EXPORT PropertyGraph {
+      return rdg_.node_entity_type_id_dictionary();
+  }
+  
+  void set_node_entity_type_id_dictionary(std::unordered_map<EntityTypeID, std::vector<EntityTypeID>>&&
```
I have special hatred for setters. IMO, every setter could have a better name. In this case, it seems the true intention is something like `persist_node_entity_type...`.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -383,13 +436,14 @@ class KATANA_EXPORT PropertyGraph {
   /// \param node_properties The node properties to copy.
   /// \param edge_properties The edge properties to copy.
-  /// \return A copy of this with a subset of the properties. The copy shares no
+  /// \return A copy of this with a subset of the properties. The copsy shares no
```
accidental change?

amberhassaan replied:
I was referring to the typo in comment above 'copsy'

amberhassaan commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -63,41 +68,100 @@ class KATANA_EXPORT RDG {
+      const std::unordered_map<EntityTypeID, std::vector<std::string>>&
+          edge_entity_type_id_name) {
+    return Store(
+        handle, command_line, IncrementVersion, std::move(topology_ff),
```
Should we really be incrementing version upon storing these arrays? @tylershunt . My understanding was that version should increment upon changes to the graph.

e-mcginnis replied:
the different use cases of the different overloaded `Store` functions isn't very clear to me, clarification of which ones should/shouldn't include these 6 structures would be be welcome. 

its not clear to me why overloading is preferable here in general vs writing wrapper functions with better defined names/tasks. That said, I'm also new to this codebase so take my opinion with a grain of salt :)

tylershunt replied:
Yeah the overloading here is unfortunate, not sure why we went with that. What's going on here (i think) is that we have two cases when we would write out an RDG:
  1. it's a new logical version of the graph so the version number needs to be bumped
  2. it's a a different view of the same logical graph so the version number should remain the same (and view info changes).
  
Would prefer just to have a `Store` and `StoreView` function rather than this cryptic enum. But I digress.

I think our changes here should be agnostic to this, in answer to @amberhassaan's question we increment the version upon write; in-memory updates don't get version numbers.

amberhassaan commented on libtsuba/src/RDGPartHeader.cpp:
```diff
@@ -191,6 +258,27 @@ tsuba::from_json(const json& j, tsuba::RDGPartHeader& header) {
+
+  if (header.storage_format_version_ >=
+      RDGPartHeader::kPartitionStorageFormatVersion2) {
+    j.at(kNodeEntityTypeIDArrayPathKey)
```
This `get_to` API is weird. Anyone know why it does that and not rerturn by value instead?

tylershunt replied:
There is a flavor that returns by value:
`template<typename T> T get()`
But it's more verbose since you can't take advantage of type inference like you can in the `get_to` case.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -237,6 +237,43 @@ class KATANA_EXPORT PropertyGraph {
     rdg_.set_view_name(fmt::format("rdg-{}-part{}", part_policy, num_hosts));
   }
 
+  const std::unordered_map<tsuba::EntityTypeID, std::vector<tsuba::EntityTypeID>>& node_entity_type_id_dictionary()
```
Why do we need this getter?

amberhassaan replied:
OK, I see that it's private. So that's good. We need a better name perhaps to signal that this is the info coming from RDG/Storage layer.

GioLomia replied:
@e-mcginnis

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -359,16 +396,32 @@ class KATANA_EXPORT PropertyGraph {
         file_(std::move(rdg_file)),
-        topology_(std::move(topo)) {}
+        topology_(std::move(topo)),
+        node_type_set_id_(std::move(node_type_ids)),
```
What about the maps, would those be added too?

GioLomia replied:
Actually, the maps will be passed in through the `rdg_` .

amberhassaan commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -63,41 +68,100 @@ class KATANA_EXPORT RDG {
+      std::unique_ptr<FileFrame> topology_ff,
+      std::unique_ptr<FileFrame> node_entity_type_id_array_ff,
+      std::unique_ptr<FileFrame> edge_entity_type_id_array_ff,
+      const std::unordered_map<EntityTypeID, std::vector<EntityTypeID>>&
```
I'm going to push back against maps and vectors as the interface between PropertyGraph and RDG. I like that arrays are being passed as FileFrames. That's good. And that we don't create a FileFrame from NUMAArray directly. The idea being that data structure choices are contained. 

Once we expose the data structure choices, a decision to change internal data structure is no longer contained to the class or even layer (logically, layers being in-memory representation and for-storage representation). 

In short, can we pass maps as something generic. For example iterator range over pairs. You can construct a map from those pairs at either end.

e-mcginnis replied:
the libtsuba side of this requested change has been made

amberhassaan reviewed (commented):
I have one high-level comment and I'll invite @scober and @roshandathathri to think about and comment. 

Seeing this PR makes me realize that we should formalize the entity type data as a class. All the necessary maps and vectors. Right now, we have a mini API explosion happening because  there is an array and 2 maps for both nodes and edges, so 6 things that need to be tracked. IMO, we should encapsulate this into a class that just deals with type IDs and APIs. 

Secondly, we should define the interface between storage and in-memory in a way that doesn't expose the internal data structure choices and doesn't bind us or corner us to stick with vector and map. 

Overall, @e-mcginnis @GioLomia , this is a good start.

amberhassaan commented:
Another general comment, the PR has 91 commits. This is good. As they say "commit early and often". However, you should practice `git rebase -i` to squash small commits. This is useful especially before you `rebase` on `master`. Right now, I see commits from other people in there, so it seems like another rebase is needed? Not sure what's going on. 

Also, since 2 of you are working together, you should do `rebase` between yourselves as well to keep a linear history and squash (rebase -i). when commits become too many.

roshandathathri commented:
> Seeing this PR makes me realize that we should formalize the entity type data as a class. All the necessary maps and vectors. Right now, we have a mini API explosion happening because there is an array and 2 maps for both nodes and edges, so 6 things that need to be tracked. IMO, we should encapsulate this into a class that just deals with type IDs and APIs.

Quick comment. Yes, that's my next action item. I'll move out everything required for entity types into a class.

e-mcginnis commented:
> Another general comment, the PR has 91 commits. This is good. As they say "commit early and often". However, you should practice `git rebase -i` to squash small commits. This is useful especially before you `rebase` on `master`. Right now, I see commits from other people in there, so it seems like another rebase is needed? Not sure what's going on.
> 
> Also, since 2 of you are working together, you should do `rebase` between yourselves as well to keep a linear history and squash (rebase -i). when commits become too many.

@GioLomia could you take a look why those commits got pulled in with the rebase?

GioLomia commented:
> > Another general comment, the PR has 91 commits. This is good. As they say "commit early and often". However, you should practice `git rebase -i` to squash small commits. This is useful especially before you `rebase` on `master`. Right now, I see commits from other people in there, so it seems like another rebase is needed? Not sure what's going on.
> > Also, since 2 of you are working together, you should do `rebase` between yourselves as well to keep a linear history and squash (rebase -i). when commits become too many.
> 
> @GioLomia could you take a look why those commits got pulled in with the rebase?

Which commits are you referring to?

roshandathathri commented:
Difficult to find the changes in this PR. Please rebase/merge master to this branch and request review again.

e-mcginnis commented:
Alright, fixed up the merge issues and did some initial squashing to cut down the noise.

roshandathathri commented:
Thanks, will make a pass now

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -269,6 +269,56 @@ class KATANA_EXPORT PropertyGraph {
 
+  const std::unordered_map<
+      tsuba::EntityTypeID, std::vector<tsuba::EntityTypeID>>&
+  node_entity_type_id_dictionary() const {
```
This is under-specified. Is the value a set of super-types or sub-types?

Define a type called `tsuba::EntityTypeIDToSetOfEntityTypeIDsMap`. I mean use the same name as that for `PropertyGraph` to make it clear that the abstraction is the same - only the representation is different. Then we can provide a way to convert between those representations. Ditto for the rest of the code.

Why is `tsuba::EntityTypeID` different from `PropertyGraph::EntityTypeID`?

e-mcginnis replied:
`tsuba::EntityTypeID` is 4 bytes while `PropertyGraph::EntityTypeID` is 1 byte. This way we can keep storage consistent, but still easily bump up memory when we reach an issue.

roshandathathri replied:
Ok, that's a good reason. Thanks for justifying. Please document that in the code as well.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -659,6 +721,26 @@ class KATANA_EXPORT PropertyGraph {
+    
+  size_t GetNodeTypeSetIDArraySize() const { return node_type_set_id_.size(); }
+
+  size_t GetEdgeTypeSetIDArraySize() const { return edge_type_set_id_.size(); }
```
Why are these two required? They should always be the number of nodes and edges in the graph respectively.

GioLomia replied:
Those functions are there to validate that assumption.

roshandathathri replied:
That should be validated when it is constructed. It shouldn't need to be exposed in the API.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -659,6 +721,26 @@ class KATANA_EXPORT PropertyGraph {
+  }
+
+  /// \return returns the node TypeSetID for @param node
+  TypeSetID GetNodeTypeSetID(Node node) const {
```
This is a duplicate of `GetTypeOfNode`

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -1551,6 +1551,6 @@ class KATANA_EXPORT PropertyGraph {
   };
 
-}  // namespace katana
+};  // namespace katana
```
; after namespace closing is not required.

e-mcginnis commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -721,836 +717,839 @@ class KATANA_EXPORT PropertyGraph {
-  bool DoesNodeHaveType(Node node, EntityTypeID node_entity_type_id) const {
-    return IsNodeSubtypeOf(node_entity_type_id, GetTypeOfNode(node));
-  }
+    /// \return true iff the node @param node has the given entity type
```
@GioLomia  not sure some of these changes were intended?

e-mcginnis replied:
alright, did some rearranging and squashing and got that handled

e-mcginnis commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -721,836 +717,839 @@ class KATANA_EXPORT PropertyGraph {
+        return ErrorCode::PropertyNotFound;
+      }
 
-  const GraphTopology& topology() const noexcept { return topology_; }
```
and this?

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -263,6 +263,69 @@ class KATANA_EXPORT PropertyGraph {
+      EntityTypeID current_id = static_cast<EntityTypeID>(pair.first);
+      EntityTypeID current_value = static_cast<EntityTypeID>(pair.second);
+
+      if (current_id >= new_entity_type_id) {
```
you could:
```
map_to_construct.resize(current_id+1);
map_to_construct[current_id].set(curent_value)
```

In fact your logic will break if `current_id` is bigger than `new_entity_type_id` by more than 1.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -263,6 +263,69 @@ class KATANA_EXPORT PropertyGraph {
+    EntityTypeIDToSetOfEntityTypeIDsMap map_to_construct;
+    for (auto& pair : iterator_range) {
+      EntityTypeID new_entity_type_id = map_to_construct.size();
+      EntityTypeID current_id = static_cast<EntityTypeID>(pair.first);
```
I'd like more meaningful names. Both key and value are EntityTypeIDs. Are they both atomic types? Seems like you are construction a mapping from Atomic to Intersection types.

GioLomia replied:
This is a mapping from EntityID to EntityID.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -263,6 +263,69 @@ class KATANA_EXPORT PropertyGraph {
+    tsuba::EntityTypeID id_key = 0;
+
+    for (const auto& type_set : id_map) {
+      SetOfEntityTypeIDs btset = type_set;
```
no need to copy.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -263,6 +263,69 @@ class KATANA_EXPORT PropertyGraph {
+
+    for (const auto& type_set : id_map) {
+      SetOfEntityTypeIDs btset = type_set;
+      for (tsuba::EntityTypeID i = 0; i < btset.size(); i++) {
```
I suspect this logic is wrong. You need to be iterating over the contents of the bitset (i.e. bits that are set). Our std::bitset is created with a fixed size of 256, so if I understand it correctly, you'll be running this loop 256 times. 

I wish bitset had an iterator over set bits. But it doesn't. So you can get around that by testing each fo the 256 bits and emitting a pair for set bits only.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -263,6 +263,69 @@ class KATANA_EXPORT PropertyGraph {
+  }
+
+  katana::Result<EntityTypeIDToAtomicTypeNameMap>
+  InflateTypeIDToNameMapFromRange(
```
Name should say that it's from atomic IDs to names.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -276,6 +339,28 @@ class KATANA_EXPORT PropertyGraph {
     rdg_.set_view_name(fmt::format("rdg-{}-part{}", part_policy, num_hosts));
   }
 
+  const EntityTypeIDToSetOfEntityTypeIDsMap& node_entity_type_id_dictionary()
```
This is bad. You're returning a reference to a temporary. It's called a dangling reference. Never do that. 
Why aren't we initializing some member of the class with we "inflate" these maps. That's the goal eventually.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -276,6 +339,28 @@ class KATANA_EXPORT PropertyGraph {
+        .value();
+  }
+
+  const EntityTypeIDToSetOfEntityTypeIDsMap& edge_entity_type_id_dictionary()
```
Same applies to these other 3 functions.

amberhassaan commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -147,6 +150,60 @@ struct PropertyColumn {
+katana::Result<std::unique_ptr<tsuba::FileFrame>>
+WriteEntityTypeIDsArray(
+    const katana::PropertyGraph::EntityTypeID* type_set_ID_array,
+    const uint32_t type_ID_array_size) {
```
NUMAArray's size type is `size_t` or you can use `uint64_t`. I prefer `size_t` as that's consistent with STL

amberhassaan commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -147,6 +150,60 @@ struct PropertyColumn {
+    return res.error();
+  }
+
+  uint32_t data[1] = {
```
Should be `uint64_t`

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -398,12 +483,24 @@ class KATANA_EXPORT PropertyGraph {
-        topology_(std::move(topo)) {}
+        topology_(std::move(topo)),
+        node_entity_type_id_(std::move(node_type_ids)),
+        edge_entity_type_id_(std::move(edge_type_ids)) {}
```
This constructor needs args for dictionaries, no?

amberhassaan commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -500,7 +608,7 @@ katana::PropertyGraph::ConstructEntityTypeIDs() {
   if (num_node_rows == 0) {
     node_entity_type_id_ = GetUnknownEntityTypeIDs(num_nodes());
-  } else {
+  } else {  ///TODO: Giorgi <---------------Not sure if this is the right way to parameterize the call for the GetEntityTypeIDsFromProperties()
```
Not sure why you need to modify this function as this is for when storage doesn't have type info.

amberhassaan commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -539,16 +647,44 @@ katana::Result<void>
+      handle, command_line, versioning_action, std::move(topology_res),
+      std::move(node_entity_type_id_array_res),
+      std::move(edge_entity_type_id_array_res),
+      FlattenEntitySetOfIDsMap(node_entity_type_id_to_atomic_entity_type_ids_),
```
This is wrong because `Flatten***` returns a range of iterators to a local variable, which will get destroyed when you return from the Flatten call. 

Flatten should instead return a vector of pairs, whose iterator range you can then pass to `rdg_.Store()`.

amberhassaan commented on libtsuba/include/tsuba/Range.h:
```diff
@@ -0,0 +1,50 @@
+ * Implements std::ranges::sized_range (C++20).
+ */
+template <typename Iterator>
+class StandardRange {
```
Please add a TODO to move this class to `libsupport` and `katana` namespace eventually.

amberhassaan commented on libtsuba/src/RDGPartHeader.cpp:
```diff
@@ -13,12 +13,27 @@ using json = nlohmann::json;
 const char* kEdgePropertyKey = "kg.v1.edge_property";
 const char* kPartPropertyFilesKey = "kg.v1.part_property_files";
 const char* kPartProperyMetaKey = "kg.v1.part_property_meta";
+const char* kStorageFormatVersionKey = "kg.v1.storage_format_version";
```
We should be declaring all of these `const char* const`

amberhassaan commented on libtsuba/src/RDGPartHeader.h:
```diff
@@ -253,6 +270,53 @@ class KATANA_EXPORT RDGPartHeader {
+    return node_entity_type_id_dictionary_;
+  }
+
+  void set_node_entity_type_id_dictionary(
```
These setters do not make sense to me. Who's calling these, as in what is the use case?

e-mcginnis replied:
the next layer up, RDG.h calls them
ex: `set_node_entity_type_id_name` in RDG.cpp
```
void
tsuba::RDG::set_node_entity_type_id_name(
    EntityTypeIDNamePairRange node_entity_type_id_name_range) {
  core_->part_header().set_node_entity_type_id_name(
      InflateEntityTypeIDNamePairRange(node_entity_type_id_name_range));
}
```

amberhassaan commented on libtsuba/src/RDG.cpp:
```diff
@@ -742,6 +951,58 @@ tsuba::RDG::set_part_metadata(const tsuba::PartitionMetadata& metadata) {
+
+tsuba::EntityTypeIDPairRange
+tsuba::RDG::edge_entity_type_id_dictionary() const {
+  return FlattenEntityTypeIDToSetOfEntitySetIDsMap(
```
Where are these inflate and Flatten methods defined?

e-mcginnis replied:
RDG.h line 367

amberhassaan reviewed (commented):
@e-mcginnis @GioLomia : Seems like you guys are return iterator ranges to local variables which get destroyed before the range can be consumed. 
Due to lack of time, you can switch to exchanging a vector of pairs if that simplifies life. Fundamentally, the range needs to point to a live variable, and that isn't happening right now.

e-mcginnis commented:
> @e-mcginnis @GioLomia : Seems like you guys are return iterator ranges to local variables which get destroyed before the range can be consumed.
> Due to lack of time, you can switch to exchanging a vector of pairs if that simplifies life. Fundamentally, the range needs to point to a live variable, and that isn't happening right now.

@amberhassaan as an un-elegant fix for this, I declared the vectors in the RDG class and still pass a range. Let me know if this is still worse than just passing the vectors themselves.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -246,10 +246,25 @@ class KATANA_EXPORT PropertyGraph {
   EntityTypeIDToAtomicTypeNameMap edge_atomic_entity_type_id_to_type_name_;
 
+  /// A vector of pairs representing flattened mapping between the node entity ids
+  tsuba::EntityTypeIDToEntityTypeIDPairs
```
I'm assuming you guys will remove these pair vectors from being members and make them temporary variables instead? We don't need to keep them around.

amberhassaan reviewed (commented):
ok, please ping me when you guys have removed the members that store pairs of IDs (and the resulting cleanup). 
Please document as well as to what the left and right member of a pair are. My understanding is that they are both atomic type IDs (or names of atomic types for strings).

GioLomia commented:
@amberhassaan I removed those unnecessary pairs. We have switched to passing vectors around. I believe the dangling reference issue should be solved as well (I might be missing some details with katana::Result).

amberhassaan commented:
@GioLomia : You've pushed other people's commits onto here. That can only happen if you didn't rebase correctly. @e-mcginnis : do you know how to fix this? If yes, please show it to @GioLomia . Thanks.

GioLomia commented:
FYI: The addition of the entity type manager will require some changes on my end. I think it would make sense to move the manager so that libtsuba can access it too. At the moment I have to unpack those values and pass them into tsuba, (I think this goes against the spirit of the abstraction). There is a fix me and @e-mcginnis have thought of, but as I said it is not the most elegant solution. The best way would be to pass the entity type manager between tsuba and PG, but right now we can not do that. @amberhassaan @roshandathathri Thoughts?

scober commented:
I need similar logic in `RDGSlice`. Could somebody ping when this logic is stable enough for me to copy over into `RDGSlice`?

GioLomia commented:
> I need similar logic in `RDGSlice`.
@scober Which logic are you referring to?

scober commented:
> > I need similar logic in `RDGSlice`.
> > @scober Which logic are you referring to?

Sorry I wasn't clear. I need to be able to load the type arrays and have access to them in memory.

amberhassaan commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -231,8 +231,7 @@ class KATANA_EXPORT PropertyGraph {
   //    for which host_i has a mirror
-  const std::vector<std::shared_ptr<arrow::ChunkedArray>>& master_nodes()
-      const {
+  const std::vector<std::shared_ptr<arrow::ChunkedArray>>& master_nodes() {
```
@scober : do you happen to know who made this function non-const?

scober replied:
I do not. Does git blame me?

amberhassaan replied:
not sure who removed the const. Wanted to make sure you didn't have a reason in particular.

amberhassaan commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -346,13 +405,85 @@ katana::GraphTopology::Copy(const GraphTopology& that) noexcept {
 katana::Result<std::unique_ptr<katana::PropertyGraph>>
 katana::PropertyGraph::Make(
     std::unique_ptr<tsuba::RDGFile> rdg_file, tsuba::RDG&& rdg) {
+  katana::GraphTopology topo;
```
This style of declaring all local variables at the top is bad. It's typically seen in C and Javascript programs. Sounds like a good idea but actually is bad for C++. Declare the variable where you need it and, more importantly, initialize it right away, unless there's a reason you can't.

amberhassaan commented on libtsuba/src/RDGPartHeader.h:
```diff
@@ -253,6 +271,159 @@ class KATANA_EXPORT RDGPartHeader {
+
+    // convert Atomic EntityType ID -> name map
+    for (const auto& pair_ : id_name) {
+      if (pair_.first <= std::numeric_limits<tsuba::EntityTypeID>::max()) {
```
this should be `std::numeric_limits<katana::EntityTypeID>...`

scober replied:
Also consider using `boost::math::tools::max_value<>` instead of `std::numeric_limits<>`. The former is defined only for known numeric types while the latter is default-defined for any type to be the default-constructed value for that type. So if we ever change the type of `EntityTypeID` to something non-numeric (like `OpaqueID`), it will be safer to use the boost option.

amberhassaan commented on libtsuba/src/RDGPartHeader.h:
```diff
@@ -253,6 +271,159 @@ class KATANA_EXPORT RDGPartHeader {
+  katana::Result<katana::EntityTypeManager> GetEntityTypeManager(
+      const tsuba::EntityTypeIDToSetOfEntityTypeIDsMap& id_dict,
+      const tsuba::EntityTypeIDToAtomicTypeNameMap& id_name) {
+    katana::EntityTypeIDToAtomicTypeNameMap manager_name_map_;
```
trailing _ is only used for class members. Please don't use it for local variables.

amberhassaan commented on libtsuba/include/tsuba/tsuba.h:
```diff
@@ -27,7 +30,10 @@ class KATANA_EXPORT RDGFile {
 
-  RDGFile(RDGFile&& f) noexcept : handle_(f.handle_) {}
+  RDGFile(RDGFile&& f) noexcept : handle_(f.handle_) {
+    f.handle_ = RDGHandle{nullptr};
```
@tylershunt : FYI: I caught a double-deletion bug in this move constructor and fixed it. Seemingly, this wasn't being called before.

e-mcginnis commented on python/katana/cpp/libgalois/graphs/Graph.pxd:
```diff
@@ -132,11 +132,14 @@ cdef extern from "katana/Graph.h" namespace "katana" nogil:
     cppclass _PropertyGraph "katana::PropertyGraph":
         PropertyGraph()
-        PropertyGraph(GraphTopology&&)
+        # PropertyGraph(GraphTopology&&)
```
@amberhassaan this seems like a mis-commit?

amberhassaan replied:
no, I intend to remove it eventually. It's replaced by the `Make` function because this constructor no longer exists in C++.

roshandathathri commented on libgalois/include/katana/PropertyGraph.h:
```diff
@@ -53,6 +53,8 @@ class KATANA_EXPORT PropertyGraph {
   using Node = GraphTopology::Node;
   using Edge = GraphTopology::Edge;
 
+  using EntityTypeIDVec = katana::NUMAArray<EntityTypeID>;
```
Don't rename `Array` to `Vec`. They have different semantics (arrays are not resizeable). I think `EntityTypeArray` is sufficient. `EntityTypeIDArray` is fine too.

e-mcginnis replied:
@amberhassaan whats your take on this?

amberhassaan commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -290,15 +291,10 @@ class EntityTypeManager {
-        SetOfEntityTypeIDs());  // for kUnknownEntityType
-    atomic_entity_type_id_to_entity_type_ids_.emplace_back(
-        SetOfEntityTypeIDs());  // for kUnknownEntityType
+    AddAtomicEntityType(kUnknownEntityTypeName);
```
also, add an assert that the returned id == kUnknownEntityType

amberhassaan commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -16,6 +16,7 @@ namespace katana {
 /// EntityTypeID is represented using 8 bits
 using EntityTypeID = uint8_t;
 static constexpr EntityTypeID kUnknownEntityType = EntityTypeID{0};
+static std::string kUnknownEntityTypeName = "kUnknownName";
```
constexpr as I said earlier. (putting it here in 1 place).

roshandathathri commented on libgalois/test/property-file-graph.cpp:
```diff
@@ -24,6 +24,94 @@ MakeProps(const std::string& name, size_t size) {
+  // KATANA_LOG_VASSERT(
+  //     g->GetNumNodeEntityTypes() == 1, "found {} entity types.",
+  //     g->GetNumNodeEntityTypes());
+  // KATANA_LOG_ASSERT(g->GetNumEdgeEntityTypes() == 1);
```
Drop commented code throughout

roshandathathri commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -52,6 +54,10 @@ class KATANA_EXPORT RDG {
 
+  /// Determine if the EntityTypeIDs are stored in properties, or outside
+  /// in their own dedicated structures
+  bool EntityTypeIDsOutsideProperties() const;
```
nit: prefer `Is<Something>` instead of `Something`

roshandathathri commented on libtsuba/include/tsuba/RDG.h:
```diff
@@ -150,6 +183,22 @@ class KATANA_EXPORT RDG {
+  /// Inform this RDG that it's Edge Entity Type Id Array is in storage at this location
+  /// without loading it into memory. \param new_type_id_array must exist and be in
+  /// the correct directory for this RDG
+  katana::Result<void> SetEdgeEntityTypeIdArrayFile(
```
nit: `Id` vs `ID` in function names can be confusing. Why not be consistent?

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -309,74 +382,101 @@ GetEntityTypeIDsFromProperties(
+  } else {
+    // we must construct id_arrays and managers from properties
+
+    KATANA_LOG_DEBUG("loading EntityType data from properties");
```
Why not warn users to update the RDG on storage?

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -448,33 +564,25 @@ katana::PropertyGraph::Validate() {
+/// Converts all uint8/bool properties into EntityTypeIDs
+/// Only call this if every uint8/bool property should be considered a type
 katana::Result<void>
 katana::PropertyGraph::ConstructEntityTypeIDs() {
```
Why is this not deleted? Where is this used?

GioLomia replied:
This is used when loading an old graph, for the sake of backwards compatibility.

e-mcginnis replied:
The enterprise repo is the only user of this now

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -448,33 +564,25 @@ katana::PropertyGraph::Validate() {
-    node_entity_type_id_ = std::move(node_types_res.value());
+  // only relevant to actually construct when EntityTypeIDs are expected in properties
+  // when EntityTypeIDs are not expected in properties then we have nothing to do here
+  if (rdg_.EntityTypeIDsOutsideProperties()) {
```
This seems wrong. The initial file loaded from storage might be version 1 but this might be called after we update the graph in memory.

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -548,11 +679,38 @@ katana::PropertyGraph::WriteView(const std::string& command_line) {
+// if we then loaded both graphs (1) and (2) in "new" software and compared them, their type information would look identical
+// but their properties information would differ as the old software added the type information to properties while the new
+// software did not. The two graphs would be functionally Equal, but this function would say this are not equal
+// TODO(unknown):(emcginnis) consider breaking the function down into: topology comparison, type comparison, and property comparison. Move pitfall described above alone with the property comparison function
```
Please move these comments into the header as doxygen `///` comments.

roshandathathri commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -578,6 +736,7 @@ katana::PropertyGraph::Equals(const PropertyGraph* other) const {
   return true;
 }
 
+/// std::string katana::PropertyGraph::ReportDiff() IS A TESTING ONLY FUNCTION, DO NOT EXPOSE THIS TO THE USER
```
Ditto

roshandathathri reviewed (commented):
All comments are nits except those related to `ConstructEntityTypeIDs`

arthurp commented on libsupport/include/katana/EntityTypeManager.h:
```diff
@@ -194,15 +198,106 @@ class EntityTypeManager {
     return (sub_atomic_types & super_atomic_types) == sub_atomic_types;
   }
 
+  const EntityTypeIDToSetOfEntityTypeIDsMap&
+  GetEntityTypeIDToAtomicEntityTypeIDs() const {
+    return entity_type_id_to_atomic_entity_type_ids_;
+  }
+
+  const EntityTypeIDToAtomicTypeNameMap& GetEntityTypeIDToAtomicTypeNameMap()
+      const {
+    return atomic_entity_type_id_to_type_name_;
+  }
```
These should NOT be exposed. The leak representation information.

e-mcginnis replied:
we have to be able to pull the information out to convert a manager to what we store. We could instead have EntityTypeManager do the conversion, but I'm not sure that is appropriate since EntityTypeManager shouldn't have to care about how we store its data.

arthurp replied:
OK. I see the issue. I'll look at it later. This is one case where I think friend classse really do make sense.

e-mcginnis replied:
definitely open to suggestions for better ways to do this in a future PR

tylershunt commented on libtsuba/src/RDG.cpp:
```diff
@@ -458,8 +619,12 @@ tsuba::RDG::DoMake(
 
-  KATANA_CHECKED(core_->EnsureNodeTypesLoaded(rdg_dir_));
-  KATANA_CHECKED(core_->EnsureEdgeTypesLoaded(rdg_dir_));
+  if (!core_->part_header().IsEntityTypeIDsOutsideProperties()) {
```
`Ensure{Node,Edge}TypesLoaded` are still required here; they are about loading the types of unloaded properties. an unfortunate overloading of our new language about node and edge types 😅 .

roshandathathri replied:
Can we please rename them to `EnsureTypesOf{Node,Edge}PropertiesLoaded`?

e-mcginnis committed:
Always need to check property types are loaded

amberhassaan commented on libgalois/src/PropertyGraph.cpp:
```diff
@@ -483,16 +586,39 @@ katana::Result<void>
 katana::PropertyGraph::DoWrite(
     tsuba::RDGHandle handle, const std::string& command_line,
     tsuba::RDG::RDGVersioningPolicy versioning_action) {
+  std::unique_ptr<tsuba::FileFrame> topology_res = nullptr;
```
The logic in this function needs to improve. First, you shouldn't be declaring variables at the top as null, because that opens up a possibility for mess ups. For example, if one of the if-conditions below is false, you'll be passing a null ptr to rdg_.Store(). can rdg_.Store() work with nullptrs?

One way to rewrite this would be:
```
if (rdg_.A.Valid() || rdg_.B.Valid() || rdg_.C.Valid()) {
    // retrun error; I'm not sure if we should return error in this case.
}
std::unique_ptr<tsuba::FileFrame> topo_res = KATANA_CHECKED(WriteTopology(topology()));
....
```

roshandathathri reviewed (approved):
LGTM. I thought Amber had already approved this.

amberhassaan commented on libtsuba/src/RDG.cpp:
```diff
@@ -267,25 +272,165 @@ tsuba::RDG::WritePartArrays(const katana::Uri& dir, tsuba::WriteGroup* desc) {
+}
+
+katana::Result<void>
+tsuba::RDG::DoStoreNodeEntityTypeIDArray(
```
These two functions (Node and Edge) version are pretty much doing the same thing. Can we combine them into 1 function in the future?

amberhassaan reviewed (approved):
some changes that can be addressed in a subsequent PR
